File: 240109-search-sort/2-15-bubble-sort.c
#include <stdio.h>
#include <stdbool.h>
 
void bubbleSort(int arr[], int n) 
{ 
   int i, j; 
   int count = 0;
   bool needIteration = true;
   while(needIteration) {
       needIteration = false;
       for(i = 0; i < n-1; i++)
       {
           for(j = 0; j < n-i-1; j++)
           {
               if(arr[j] > arr[j+1])
               { 
                   int temp = arr[j]; 
                   arr[j] = arr[j+1]; 
                   arr[j+1] = temp; 
                   needIteration = true;
                   count++;
                }
            }
        }
    }
   printf("The number of permutations is %d. \n", count); 
} 
 
void showArray(int arr[], int size) 
{ 
    int i; 
    for(i=0; i < size; i++) 
        printf("%d ", arr[i]); 
    printf("\n"); 
}
 
int main()
{ 
    int libraryNum[] = {124,235,456,123,756,476,285,998,379,108}; 
    int n = sizeof(libraryNum)/sizeof(libraryNum[0]); 
 
    printf("Initial array: \n"); 
    showArray(libraryNum, n); 
    bubbleSort(libraryNum, n); 
    printf("Sorted array: \n"); 
    showArray(libraryNum, n); 
    return 0; 
}

File: 240109-search-sort/2-15-bubble-sort.py
def bubble_sort(arr):
    count = 0
    need_iteration = 'true'
    while need_iteration == 'true':
        need_iteration = 'false'
        for i in range(len(arr)):
            for j in range(0, len(arr)-i-1):
                if arr[j] > arr[j+1]:
                    arr[j], arr[j+1] = arr[j+1], arr[j]
                    need_iteration = 'true'
                    count +=1
    print("The number of permutations is", count)
 
libraryNum = [124,235,456,123,756,476,285,998,379,108] 
print("Initial array:", libraryNum) 
bubble_sort(libraryNum) 
print ("Sorted array:", libraryNum)